mackerel-plugin-aws-elasticsearch/lib/aws-elasticsearch.go

```diff
@@ -14,26 +13,50 @@ import (
 	mp "github.com/mackerelio/go-mackerel-plugin"
 )
 
+const (
+	nameSpace          = "AWS/ES"
+	metricsTypeAverage = "Average"
+	metricsTypeSum     = "Sum"
+	metricsTypeMaximum = "Maximum"
+	metricsTypeMinimum = "Minimum"
+)
+
 var graphdef = map[string]mp.Graphs{
+	"es.ClusterStatus": {
+		Label: "AWS ES ClusterStatus",
+		Unit:  "integer",
+		Metrics: []mp.Metrics{
+			{Name: "ClusterStatus.green", Label: "green"},
+			{Name: "ClusterStatus.yellow", Label: "yellow"},
+			{Name: "ClusterStatus.red", Label: "red"},
```

Comment:
Basically, we should not use a dot here. The graph title will not be applied correctly. With the current implementation, metric name is `custom.es.ClusterStatus. ClusterStatus.green` and the graph title is `custom.es.ClusterStatus.ClusterStatus*`. An easy solution for this is to change the metric names to snake_case (or kebab-case or whatever) and replace the separator with dot just before collecting the metrics from AWS.

Reply:
As reference, the document ([link](https://mackerel.io/docs/entry/advanced/custom-metrics), [ja](https://mackerel.io/ja/docs/entry/advanced/custom-metrics)) mentions that the metric name in the graph definition should match `[-a-zA-Z0-9_]+`. The creating graph API will respond OK even if the body contains an invalid definition (for backwards compatibility), but the response body should contain the names of invalid graph definitions. It's still hard for plugin creator to tell what's going on for invalid definitions and automatically created graphs so we'll continue to improve the documentation and the behavior of the graph api.

Reply:
I tried with the code (09e7c0a82ed6e1dba14d2d475ce51fe15376cfa4) before I modified it. 

```
$ pwd
/Users/hidekazu/src/github.com/mackerelio/mackerel-agent-plugins/mackerel-plugin-aws-elasticsearch
$ git checkout 09e7c0a82ed6e1dba14d2d475ce51fe15376cfa4
$ env MACKEREL_AGENT_PLUGIN_META=1 go run main.go -domain=xxxx -client-id=xxxx -region=ap-northeast-1 | tail -n 1 |jq '.'
{
  "graphs": {
    "es.AutomatedSnapshotFailure": {
      "label": "AWS ES AutomatedSnapshotFailure",
      "unit": "integer",
      "metrics": [
        {
          "name": "AutomatedSnapshotFailure",
          "label": "AutomatedSnapshotFailure",
          "stacked": false
        }
      ]
    },
    "es.CPUUtilization": {
      "label": "AWS ES CPU Utilization",
      "unit": "percentage",
      "metrics": [
        {
          "name": "CPUUtilization",
          "label": "CPUUtilization",
          "stacked": false
        }
      ]
    },
    "es.ClusterStatus": {
      "label": "AWS ES ClusterStatus",
      "unit": "integer",
      "metrics": [
        {
          "name": "ClusterStatus.green",
          "label": "green",
          "stacked": false
        },
        {
          "name": "ClusterStatus.yellow",
          "label": "yellow",
          "stacked": false
        },
        {
          "name": "ClusterStatus.red",
          "label": "red",
          "stacked": false
        }
      ]
    },
    "es.DeletedDocuments": {
      "label": "AWS ES DeletedDocuments",
      "unit": "integer",
      "metrics": [
        {
          "name": "DeletedDocuments",
          "label": "DeletedDocuments",
          "stacked": false
        }
      ]
    },
    "es.DiskQueueDepth": {
      "label": "AWS ES DiskQueueDepth",
      "unit": "integer",
      "metrics": [
        {
          "name": "DiskQueueDepth",
          "label": "DiskQueueDepth",
          "stacked": false
        }
      ]
    },
    "es.FreeStorageSpace": {
      "label": "AWS ES Free Storage Space",
      "unit": "bytes",
      "metrics": [
        {
          "name": "FreeStorageSpace",
          "label": "FreeStorageSpace",
          "stacked": false
        }
      ]
    },
    "es.IOPS": {
      "label": "AWS ES IOPS",
      "unit": "iops",
      "metrics": [
        {
          "name": "ReadIOPS",
          "label": "ReadIOPS",
          "stacked": false
        },
        {
          "name": "WriteIOPS",
          "label": "WriteIOPS",
          "stacked": false
        }
      ]
    },
    "es.JVMMemoryPressure": {
      "label": "AWS ES JVMMemoryPressure",
      "unit": "percentage",
      "metrics": [
        {
          "name": "JVMMemoryPressure",
          "label": "JVMMemoryPressure",
          "stacked": false
        }
      ]
    },
    "es.Latency": {
      "label": "AWS ES Latency",
      "unit": "float",
      "metrics": [
        {
          "name": "ReadLatency",
          "label": "ReadLatency",
          "stacked": false
        },
        {
          "name": "WriteLatency",
          "label": "WriteLatency",
          "stacked": false
        }
      ]
    },
    "es.MasterCPUUtilization": {
      "label": "AWS ES MasterCPUUtilization",
      "unit": "percentage",
      "metrics": [
        {
          "name": "MasterCPUUtilization",
          "label": "MasterCPUUtilization",
          "stacked": false
        }
      ]
    },
    "es.MasterFreeStorageSpace": {
      "label": "AWS ES MasterFreeStorageSpace",
      "unit": "bytes",
      "metrics": [
        {
          "name": "MasterFreeStorageSpace",
          "label": "MasterFreeStorageSpace",
          "stacked": false
        }
      ]
    },
    "es.MasterJVMMemoryPressure": {
      "label": "AWS ES MasterJVMMemoryPressure",
      "unit": "percentage",
      "metrics": [
        {
          "name": "MasterJVMMemoryPressure",
          "label": "MasterJVMMemoryPressure",
          "stacked": false
        }
      ]
    },
    "es.Nodes": {
      "label": "AWS ES Nodes",
      "unit": "integer",
      "metrics": [
        {
          "name": "Nodes",
          "label": "Nodes",
          "stacked": false
        }
      ]
    },
    "es.SearchableDocuments": {
      "label": "AWS ES SearchableDocuments",
      "unit": "integer",
      "metrics": [
        {
          "name": "SearchableDocuments",
          "label": "SearchableDocuments",
          "stacked": false
        }
      ]
    },
    "es.Throughput": {
      "label": "AWS ES Throughput",
      "unit": "bytes/sec",
      "metrics": [
        {
          "name": "ReadThroughput",
          "label": "ReadThroughput",
          "stacked": false
        },
        {
          "name": "WriteThroughput",
          "label": "WriteThroughput",
          "stacked": false
        }
      ]
    }
  }
}
$ go run main.go -domain=xxxx -client-id=xxxx -region=ap-northeast-1
es.MasterFreeStorageSpace.MasterFreeStorageSpace	3122520915.968000	1507854582
es.DiskQueueDepth.DiskQueueDepth	0.000183	1507854582
es.Latency.ReadLatency	0	1507854582
es.Latency.WriteLatency	0.001885	1507854582
es.JVMMemoryPressure.JVMMemoryPressure	43.700000	1507854582
es.SearchableDocuments.SearchableDocuments	7751	1507854582
es.AutomatedSnapshotFailure.AutomatedSnapshotFailure	0	1507854582
es.MasterCPUUtilization.MasterCPUUtilization	0	1507854582
es.Nodes.Nodes	5	1507854582
es.CPUUtilization.CPUUtilization	0.500000	1507854582
es.MasterJVMMemoryPressure.MasterJVMMemoryPressure	19.400000	1507854582
es.ClusterStatus.ClusterStatus.green	1	1507854582
es.ClusterStatus.ClusterStatus.yellow	0	1507854582
es.ClusterStatus.ClusterStatus.red	0	1507854582
es.FreeStorageSpace.FreeStorageSpace	63126541172.736000	1507854582
es.DeletedDocuments.DeletedDocuments	10	1507854582
es.IOPS.ReadIOPS	0	1507854582
es.IOPS.WriteIOPS	0.098333	1507854582
es.Throughput.ReadThroughput	0	1507854582
es.Throughput.WriteThroughput	1024	1507854582
```

I tried it at the current master branch.

```
$ git checkout master
$ env MACKEREL_AGENT_PLUGIN_META=1 go run main.go -domain=xxxx -client-id=xxxx -region=ap-northeast-1 | tail -n 1 |jq '.'
{
  "graphs": {
    "es.AutomatedSnapshotFailure": {
      "label": "AWS ES AutomatedSnapshotFailure",
      "unit": "integer",
      "metrics": [
        {
          "name": "AutomatedSnapshotFailure",
          "label": "AutomatedSnapshotFailure",
          "stacked": false
        }
      ]
    },
    "es.CPUUtilization": {
      "label": "AWS ES CPU Utilization",
      "unit": "percentage",
      "metrics": [
        {
          "name": "CPUUtilization",
          "label": "CPUUtilization",
          "stacked": false
        }
      ]
    },
    "es.ClusterIndexWritesBlocked": {
      "label": "AWS ES ClusterIndexWritesBlocked",
      "unit": "integer",
      "metrics": [
        {
          "name": "ClusterIndexWritesBlocked",
          "label": "ClusterIndexWritesBlocked",
          "stacked": false
        }
      ]
    },
    "es.ClusterStatus": {
      "label": "AWS ES ClusterStatus",
      "unit": "integer",
      "metrics": [
        {
          "name": "ClusterStatus.green",
          "label": "green",
          "stacked": false
        },
        {
          "name": "ClusterStatus.yellow",
          "label": "yellow",
          "stacked": false
        },
        {
          "name": "ClusterStatus.red",
          "label": "red",
          "stacked": false
        }
      ]
    },
    "es.ClusterUsedSpace": {
      "label": "AWS ES Cluster Used Space",
      "unit": "bytes",
      "metrics": [
        {
          "name": "ClusterUsedSpace",
          "label": "ClusterUsedSpace",
          "stacked": false
        }
      ]
    },
    "es.DeletedDocuments": {
      "label": "AWS ES DeletedDocuments",
      "unit": "integer",
      "metrics": [
        {
          "name": "DeletedDocuments",
          "label": "DeletedDocuments",
          "stacked": false
        }
      ]
    },
    "es.DiskQueueDepth": {
      "label": "AWS ES DiskQueueDepth",
      "unit": "integer",
      "metrics": [
        {
          "name": "DiskQueueDepth",
          "label": "DiskQueueDepth",
          "stacked": false
        }
      ]
    },
    "es.FreeStorageSpace": {
      "label": "AWS ES Free Storage Space",
      "unit": "bytes",
      "metrics": [
        {
          "name": "FreeStorageSpace",
          "label": "FreeStorageSpace",
          "stacked": false
        }
      ]
    },
    "es.IOPS": {
      "label": "AWS ES IOPS",
      "unit": "iops",
      "metrics": [
        {
          "name": "ReadIOPS",
          "label": "ReadIOPS",
          "stacked": false
        },
        {
          "name": "WriteIOPS",
          "label": "WriteIOPS",
          "stacked": false
        }
      ]
    },
    "es.JVMMemoryPressure": {
      "label": "AWS ES JVMMemoryPressure",
      "unit": "percentage",
      "metrics": [
        {
          "name": "JVMMemoryPressure",
          "label": "JVMMemoryPressure",
          "stacked": false
        }
      ]
    },
    "es.KibanaHealthyNodes": {
      "label": "AWS ES KibanaHealthyNodes",
      "unit": "integer",
      "metrics": [
        {
          "name": "KibanaHealthyNodes",
          "label": "KibanaHealthyNodes",
          "stacked": false
        }
      ]
    },
    "es.Latency": {
      "label": "AWS ES Latency",
      "unit": "float",
      "metrics": [
        {
          "name": "ReadLatency",
          "label": "ReadLatency",
          "stacked": false
        },
        {
          "name": "WriteLatency",
          "label": "WriteLatency",
          "stacked": false
        }
      ]
    },
    "es.MasterCPUUtilization": {
      "label": "AWS ES MasterCPUUtilization",
      "unit": "percentage",
      "metrics": [
        {
          "name": "MasterCPUUtilization",
          "label": "MasterCPUUtilization",
          "stacked": false
        }
      ]
    },
    "es.MasterFreeStorageSpace": {
      "label": "AWS ES MasterFreeStorageSpace",
      "unit": "bytes",
      "metrics": [
        {
          "name": "MasterFreeStorageSpace",
          "label": "MasterFreeStorageSpace",
          "stacked": false
        }
      ]
    },
    "es.MasterJVMMemoryPressure": {
      "label": "AWS ES MasterJVMMemoryPressure",
      "unit": "percentage",
      "metrics": [
        {
          "name": "MasterJVMMemoryPressure",
          "label": "MasterJVMMemoryPressure",
          "stacked": false
        }
      ]
    },
    "es.MasterReachableFromNode": {
      "label": "AWS ES MasterReachableFromNode",
      "unit": "percentage",
      "metrics": [
        {
          "name": "MasterReachableFromNode",
          "label": "MasterReachableFromNode",
          "stacked": false
        }
      ]
    },
    "es.Nodes": {
      "label": "AWS ES Nodes",
      "unit": "integer",
      "metrics": [
        {
          "name": "Nodes",
          "label": "Nodes",
          "stacked": false
        }
      ]
    },
    "es.SearchableDocuments": {
      "label": "AWS ES SearchableDocuments",
      "unit": "integer",
      "metrics": [
        {
          "name": "SearchableDocuments",
          "label": "SearchableDocuments",
          "stacked": false
        }
      ]
    },
    "es.Throughput": {
      "label": "AWS ES Throughput",
      "unit": "bytes/sec",
      "metrics": [
        {
          "name": "ReadThroughput",
          "label": "ReadThroughput",
          "stacked": false
        },
        {
          "name": "WriteThroughput",
          "label": "WriteThroughput",
          "stacked": false
        }
      ]
    }
  }
}
$ go run main.go -domain=xxxx -client-id=xxxx -region=ap-northeast-1
es.FreeStorageSpace.FreeStorageSpace	63126541172.736000	1507854729
es.SearchableDocuments.SearchableDocuments	7751	1507854729
es.ClusterUsedSpace.ClusterUsedSpace	27085550	1507854729
es.MasterFreeStorageSpace.MasterFreeStorageSpace	3122496798.720000	1507854729
es.Nodes.Nodes	5	1507854729
es.DeletedDocuments.DeletedDocuments	10	1507854729
es.JVMMemoryPressure.JVMMemoryPressure	49.300000	1507854729
es.MasterJVMMemoryPressure.MasterJVMMemoryPressure	19.400000	1507854729
es.MasterReachableFromNode.MasterReachableFromNode	1	1507854729
es.ClusterStatus.ClusterStatus.green	1	1507854729
es.ClusterStatus.ClusterStatus.yellow	0	1507854729
es.ClusterStatus.ClusterStatus.red	0	1507854729
es.AutomatedSnapshotFailure.AutomatedSnapshotFailure	0	1507854729
es.KibanaHealthyNodes.KibanaHealthyNodes	1	1507854729
es.MasterCPUUtilization.MasterCPUUtilization	0	1507854729
es.CPUUtilization.CPUUtilization	1	1507854729
```

Comparing both, the output method of ClusterStatus has not changed.
But,  it is good to fix this problem.